src/main/java/com/gtnewhorizon/gtnhlib/concurrent/cas/CasList.java

```
@@ -0,0 +1,328 @@
+public class CasList<T> extends CasAdapter<ObjectImmutableList<T>, ObjectArrayList<T>> implements List<T> {
+
+    /** Constructs an empty list */
+    public CasList() {
+        super(ObjectImmutableList.of());
+    }
+
+    /** Constructs a list from the given elements, or an empty list if null */
+    @SafeVarargs
+    public CasList(final @Nullable T... initialElements) {
+        super(initialElements == null ? ObjectImmutableList.of() : new ObjectImmutableList<>(initialElements));
+    }
+
+    /** Constructs a list from the given elements, or an empty list if null */
+    public CasList(final @Nullable Collection<T> initialElements) {
+        super(initialElements == null ? ObjectImmutableList.of() : new ObjectImmutableList<>(initialElements));
+    }
+
+    /** Constructs a list from the given elements, or an empty list if null */
+    public CasList(final @Nullable ObjectIterator<T> initialElements) {
+        super(initialElements == null ? ObjectImmutableList.of() : new ObjectImmutableList<>(initialElements));
+    }
+
+    /** Constructs a list populated in the given lambda */
+    public CasList(final @NotNull Consumer<@NotNull ObjectArrayList<T>> constructor) {
+        this();
+        final ObjectArrayList<T> elems = new ObjectArrayList<>();
+        constructor.accept(elems);
+        overwrite(immutableCopyOf(elems));
+    }
+
+    @Override
+    public final @NotNull ObjectArrayList<T> mutableCopyOf(@NotNull ObjectImmutableList<T> data) {
+        return new ObjectArrayList<>(data);
+    }
+
+    @Override
+    public final @NotNull ObjectImmutableList<T> immutableCopyOf(@NotNull ObjectArrayList<T> data) {
+        return new ObjectImmutableList<>(data);
+    }
+
+    /**
+     * @deprecated Basically useless as it could be modified after being checked, iterate over a stored reference from
+     *             {@link CasList#read()} instead.
+     */
+    @Override
+    @Deprecated
+    public int size() {
+        return read().size();
```

Should these even be deprecated or should they just throw UnsupportedOperationException?

---

I'd like for this to work as a drop-in, and theoretically it's fine when you use size() only to report the current size or do some other metrics that aren't just iterating in a for-loop